### PR TITLE
Refactor Artifacts handling.

### PR DESCRIPTION
**Pull request by andreidenissov-cog**
First step to implementing Feature/394.
Mostly refactoring artifacts handling by separating artifacts logic
into its own Artifact class.
With this PR, functionality remains basically the same,
but we put in the base on which to build intended extensions.

### CODE REVIEW

File: studio/local_worker.py [hunk 1]
```diff
@@ -77,7 +78,7 @@ def run(self, experiment):
                 python = which(python)
 
                 cmd = [python, experiment.filename] + experiment.args
-                cwd = experiment.artifacts['workspace']['local']
+                cwd = experiment.artifacts['workspace'].local_path
```
Comment (karlmutch) [inline studio/local_worker.py:81]: Does naming this var explicitly named rather in a map hurt extensibility ? Not that we would extend it much of course.

File: studio/local_worker.py [hunk 2]
```diff
@@ -388,21 +389,25 @@ def hold_job():
...
-                                    sleep_time=10,
-                                    logger=logger)
-                            else:
-                                art['local'] = retry(
-                                    lambda: db.get_artifact(art),
+                                get_only_newer = False
+
+                            if not art.is_mutable:
+                                logger.info('Fetching artifact ' + tag)
+                                print("IMMUTABLE: {0} {1}".format(tag, json.dumps(art.to_dict(), indent=4)))
```
Comment (karlmutch) [inline studio/local_worker.py:401]: Debug, should this go ?

File: studio/model.py [hunk 3]
```diff
@@ -64,66 +64,63 @@ def replace_with_env(config):
...
-    db_provider = None
     db_config = config['database']
     if db_config['type'].lower() == 'firebase':
-        db_provider = FirebaseProvider(
-            db_config,
-            blocking_auth,
-            verbose=verbose,
-            store=artifact_store)
-        artifact_store = db_provider.get_artifact_store()
+        raise NotImplementedError("Firebase is not supported.")
```
Comment (karlmutch) [inline studio/model.py:103]: Is it possible to not drop firebase.  It was being used for the model zoo.  We could resurrect it one day

File: studio/storage_handler.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import os
+import uuid
+import time
+
+from .storage_type import StorageType
+from .util import get_temp_filename
+
+class StorageHandler(object):
```
Comment (karlmutch) [inline studio/storage_handler.py:8]: Add a purpose comment.

File: studio/storage_type.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from enum import Enum
+
```
Comment (karlmutch) [inline studio/storage_type.py:2]: Is this very useful or could testing attempted casting do this for us.

**Review by karlmutch — CHANGES_REQUESTED** [review]

File: studio/experiment.py [hunk 6]
```diff
@@ -54,7 +55,7 @@ def __init__(self, key, filename, args, pythonenv,
         except BaseException:
             model_dir = None
 
-        self.artifacts = {
+        std_artifacts_dict = {
```
[the remark this answers is not in the document]
Comment (andreidenissov-cog) [reply studio/experiment.py:58]: No, this is just a local variable name (with a hint that this is actually a dictionary, not an Artifact instance)

File: studio/artifact.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+except ImportError:
+    from urllib import urlopen
+
+from . import fs_tracker, util, logs
+from . import model_setup
+from .storage_type import StorageType
+from .storage_handler import StorageHandler
+from .storage_util import tar_artifact, untar_artifact
+
+class Artifact:
```
[the remark this answers is not in the document]
Comment (andreidenissov-cog) [reply studio/artifact.py:17]: Commented in artifact.py

File: studio/experiment.py [hunk 8]
```diff
@@ -87,7 +93,19 @@ def __init__(self, key, filename, args, pythonenv,
             }
         }
         if artifacts is not None:
-            self.artifacts.update(artifacts)
+            std_artifacts_dict.update(artifacts)
+
+        # Semi-hack: make sure 'retval' artifact does have
+        # local path setup
+        # (it could be overwritten by experiment artifacts provided
+        # by constuctor parameters.)
```
[the remark this answers is not in the document]
Comment (andreidenissov-cog) [reply studio/experiment.py:101]: yep.

File: studio/local_worker.py (hunk 1 above)
Comment (andreidenissov-cog) [reply studio/local_worker.py:81]: experiment.artifacts['workspace'] here is an instance of Artifact class, so 'local_path' is a property of this instance.

File: studio/local_worker.py (hunk 2 above)
Comment (andreidenissov-cog) [reply studio/local_worker.py:401]: Yes, that's gone.

File: studio/model.py (hunk 3 above)
Comment (andreidenissov-cog) [reply studio/model.py:103]: One day, right. OK put back Firebase in the form of FirebaseStorageHandler which will provide storage operations for Firebase endpoints.
Not tested as of now.

File: studio/storage_handler.py (hunk 4 above)
Comment (andreidenissov-cog) [reply studio/storage_handler.py:8]: Done.

File: studio/storage_type.py (hunk 5 above)
Comment (andreidenissov-cog) [reply studio/storage_type.py:2]: It could for sure, but looking at simple enum value is cleaner in my opinion.

File: studio/model.py (hunk 3 above)
Comment (karlmutch) [reply studio/model.py]: Thats OK

**Review by karlmutch — APPROVED** [review]
LGTM

**Pull request merged**